lib/Sema/TypeCheckAttr.cpp

```diff
@@ -4749,6 +4749,8 @@ static bool typeCheckDerivativeAttr(ASTContext &Ctx, Decl *D,
   // to be enabled.
   if (checkIfDifferentiableProgrammingEnabled(Ctx, attr, D->getDeclContext()))
     return true;
+  if (D->isInvalid())
```

Comment:
Checking the invalid bit and skipping work in Sema is certainly *a way* to make crashes go away. But we probably want to identify a root cause and fix that first. There's very few instances where this pattern actually helps. In practice it can hinder many more processes downstream as it either hides bugs, shifts data dependencies, or hides dependencies from the incremental build. We're trying to eliminate usages of this pattern where we can - unfortunately many of them are load bearing.

Reply:
The root cause is that we're calling `setInvalid()` when a generic parameter is not used in the type signature of a function/subscript etc.

Reply:
So the real question is: why was `typeCheckDerivativeAttr` called when an error diagnostic was issued?

Is this correct?

Reply:
> why was typeCheckDerivativeAttr called when an error diagnostic was issued?

We try not to skip performing semantic analysis just because the AST was malformed. The purpose of Sema is, after all, to explain to the user what went wrong. 

> The root cause is that we're calling setInvalid() when a generic parameter is not used in the type signature of a function/subscript etc., although there seems to be nothing wrong with leaving the interface type as-is.

I agree that this seems wrong. You could probably start by just dropping the call to `setInvalid()` on the path and seeing what changes.

Reply:
> The root cause is that we're calling setInvalid() when a generic parameter is not used in the type signature of a function/subscript etc., although there seems to be nothing wrong with leaving the interface type as-is.

Agreed. Start by eliminating that call and seeing what changes.

Reply:
@AnthonyLatsis I have my mind focused on the new bug I'm to file on JIRA. Would you be able to investigate that call and send over the necessary changes via a code review?

Reply:
> Start by eliminating that call and seeing what changes.

Pretty reassuring, running tests locally shows only 2 new `generic parameter 'T' could not be inferred` errors that are now expected. 
___

@philipturner if you are willing to see this through, you have to delete [this](https://github.com/apple/swift/blob/cb27114b8853ace172e325fe47a086873b215367/lib/Sema/TypeCheckGeneric.cpp#L509) call and update the following failing tests (and add the AutoDiff regression test):

```
Swift(macosx-x86_64) :: Constraints/overload.swift
Swift(macosx-x86_64) :: Generics/unbound.swift
```

Reply:
Got it! Thanks for narrowing down the crux of the bug!

Reply:
> (and add the AutoDiff regression test):

That just means add a new test proving this bug was fixed, correct? If so, I'll need a bit of help setting that up.

Reply:
We already have a test file for `@derivative` checking: `../test/AutoDiff/Sema/derivative_attr_type_checking.swift`. Just add your test case to that and run `lit` on it to see the diagnostics that are expected. If the expectations are correct, incorporate the missing diagnostics into your test using `expected-*` directives, as shown elsewhere in that file. The lit command is

```
cd ../build/Ninja-RelWithDebInfoAssert/test-macosx-$(uname -m) && ../llvm-project/llvm/utils/lit/lit.py -sv --filter=derivative_attr_type_checking AutoDiff/Sema
```

Reply:
@AnthonyLatsis done!